### PR TITLE
Tdg 11

### PR DESCRIPTION
**Pull request by GianiStatie**
What was done:
- now allies are made out of several sprites (base, hair, weapon, front hand)
- created logic for ally pursuing enemy 
- create logic + animation for ally attacking enemy 
- added hitbox/hurtbox (you get hurt when a hibox connects with a hurtbox)
- added logic for ally to go back at spawn point once enemy is outside range
- (not yet) add ranged ally
- (not yet) ranged ally spawns projectiles that hit enemies

### CODE REVIEW

File: main.tscn [hunk 1]
```diff
@@ -1,9 +1,9 @@
 [gd_scene load_steps=22 format=3 uid="uid://c15xsnw5jwls3"]
 
-[ext_resource type="Texture2D" uid="uid://c674m2xx0bv8b" path="res://assets/roguelikeSheet_magenta.png" id="1_70rxh"]
+[ext_resource type="Texture2D" uid="uid://dyd34q7rcn5wv" path="res://assets/roguelikeSheet_magenta.png" id="1_70rxh"]
 [ext_resource type="Script" path="res://Main.gd" id="1_jhswa"]
 [ext_resource type="PackedScene" uid="uid://c035nhgeltao5" path="res://src/structures/spawner.tscn" id="2_ld6ey"]
-[ext_resource type="Texture2D" uid="uid://dhfj0mbeol4r5" path="res://assets/tilesets/Sprite-0001.png" id="2_servy"]
+[ext_resource type="Texture2D" uid="uid://do8ns62bsespb" path="res://assets/tilesets/Sprite-0001.png" id="2_servy"]
```
Comment (Ratoone) [inline main.tscn:6]: This is so annoying, isn't there a way to use the res path instead? I also get tons of warnings in the editor about it

File: src/units/00_components/Hitbox.gd [hunk 2]
```diff
@@ -0,0 +1,3 @@
+extends Area2D
+
+@export var damage = 10
```
Comment (Ratoone) [inline src/units/00_components/Hitbox.gd:3]: I think the damage should depend on the tower stat

File: src/units/02_allies/states_mercenary/AttackMercenary.gd [hunk 3]
```diff
@@ -0,0 +1,9 @@
+extends State
+class_name AttackState
+
+
+func enter(_msg := {}) -> void:
+	owner.animation_player.play("MeleAttack")
```
Comment (Ratoone) [inline src/units/02_allies/states_mercenary/AttackMercenary.gd:6]: Garmmar hard 😂

Comment (Ratoone) [reply src/units/02_allies/states_mercenary/AttackMercenary.gd:6]: Also, I wouldn't put it under the mercenary, maybe make a state in the AbstractTower called attack, and send in the msg whether it's a melee or ranged

File: src/units/02_allies/states_mercenary/IdleMercenary.gd [hunk 4]
```diff
@@ -0,0 +1,14 @@
+extends StateIdle
+class_name IdleMercenary
```
Comment (Ratoone) [inline src/units/02_allies/states_mercenary/IdleMercenary.gd:2]: Same comment as above. If melee, do the Pursue, else shoot

File: src/units/02_allies/states_mercenary/RecallMercenary.gd [hunk 5]
```diff
@@ -0,0 +1,12 @@
+extends State
+class_name RecallMercenary
```
Comment (Ratoone) [inline src/units/02_allies/states_mercenary/RecallMercenary.gd:2]: What is this used for?

Comment (GianiStatie) [reply src/units/02_allies/states_mercenary/RecallMercenary.gd:2]: Nothing, will remove it.

File: src/units/02_allies/states_mercenary/AttackMercenary.gd [hunk 6]
```diff
@@ -1,9 +1,9 @@
 extends State
-class_name AttackState
+class_name MeleeAttackMercenary
```
Comment (Ratoone) [inline src/units/02_allies/states_mercenary/AttackMercenary.gd:2]: i'm still not happy this is not a generic attack state, and is specific to the mercenary -.-

File: src/units/02_allies/states_mercenary/DeployMercenary.gd [hunk 7]
```diff
@@ -0,0 +1,16 @@
+extends State
+class_name DeplyMercenaryState
```
Comment (Ratoone) [inline src/units/02_allies/states_mercenary/DeployMercenary.gd:2]: Same here. Don't all the towers deploy the same?

File: src/units/02_allies/states_mercenary/PursueMercenary.gd [hunk 8]
```diff
@@ -1,13 +1,22 @@
 extends StateMove
 class_name PursueMercenary
```
Comment (Ratoone) [inline src/units/02_allies/states_mercenary/PursueMercenary.gd:2]: Same here

File: src/units/02_allies/states_mercenary/PursueMercenary.gd [hunk 9]
```diff
@@ -18,7 +27,21 @@ func update(delta):
 	super.update(delta)
 
 func _on_arrived_at_target():
+	print('should not get near enemy if range is long enough')
 	state_machine.transition_to("Idle")
 
-func _on_attack_range_area_entered(area):
-	state_machine.transition_to("Attack")
+func _on_attack_range_area_entered(_area):
+	if not active:
```
Comment (Ratoone) [inline src/units/02_allies/states_mercenary/PursueMercenary.gd:34]: y u not use `state_machine.state == self`? it's cleaner than using an `active` flag

File: src/units/02_allies/states_mercenary/PursueMercenary.gd [hunk 10]
```diff
@@ -18,7 +27,21 @@ func update(delta):
...
+		return
+	if owner.stats.weapon_type == "Mele":
+		state_machine.transition_to("MeleeAttack")
+	elif owner.stats.weapon_type == "Ranged":
+		state_machine.transition_to("RangedAttack", {"enemy_position": enemy.global_position})
+	else:
+		assert(false, "ERROR: Not implemented error")
```
Comment (Ratoone) [inline src/units/02_allies/states_mercenary/PursueMercenary.gd:41]: yess, perfect

File: src/units/02_allies/states_mercenary/PursueMercenary.gd [hunk 11]
```diff
@@ -18,7 +27,21 @@ func update(delta):
...
+
+func is_enemy_in_range():
+	for area in attack_range.get_overlapping_areas():
+		if enemy == area.owner:
+			return true
+	false
```
Comment (Ratoone) [inline src/units/02_allies/states_mercenary/PursueMercenary.gd:47]: uuuh, rust syntax, didn't know you can do that

**Comment by Ratoone** [comment]
After going through the whole thing, i realised that mercenary is the term you used for the "towers". My bad

File: src/units/02_allies/states_mercenary/PursueMercenary.gd (hunk 9 above)
Comment (GianiStatie) [reply src/units/02_allies/states_mercenary/PursueMercenary.gd:34]: cuz my brain no work

**Pull request merged**